src/glang/codegen/expressions.py

```diff
@@ -344,7 +381,10 @@ def __init__(
             self._indices.append(index_expr)
             self._conversion_exprs.extend(conversions)
 
-        super().__init__(self._element_type.convert_to_value_type(), True)
+        pure_type = array_ptr.get_equivalent_pure_type()
+        super().__init__(
+            self._element_type.convert_to_value_type(), pure_type.storage_kind
```

Comment:
`pure_type.storage_kind` is always a reference here (since we've asserted the array has an address).

`array_ptr.underlying_indirection_kind` may be a value since `(&a)[1]`  is a valid expression. In this case, `underlying_indirection_kind == Type.Kind.VALUE` since we don't support indirect assesses to reference types.